lib/ansible/modules/web_infrastructure/jboss.py

```diff
@@ -43,45 +43,140 @@
   - "Ensure no identically named application is deployed through the JBoss CLI"
+  - "At a minimum, url_password should be vaulted."
+  - "HTTP management API is supported in JBoss AS 7.1, Wildfly >= 8, and JBoss EAP >= 6."
+  - "JBoss 5 supports filesystem deployments only."
+  - "Filesystem deployments should be avoided in production environments where possible."
 author: "Jeroen Hoekx (@jhoekx)"
 """
 
 EXAMPLES = """
-# Deploy a hello world application
+# Deploy a hello world application using filesystem
 - jboss:
     src: /tmp/hello-1.0-SNAPSHOT.war
     deployment: hello.war
-    state: present
+    state: deployed
 
-# Update the hello world application
+# Update the hello world application using filesystem
 - jboss:
     src: /tmp/hello-1.1-SNAPSHOT.war
     deployment: hello.war
-    state: present
+    state: deployed
 
-# Undeploy the hello world application
+# Undeploy the hello world application using filesystem
 - jboss:
     deployment: hello.war
-    state: absent
+    state: undeployed
+
+# Deploy the hello world application using HTTP management API
+- jboss:
+    src: /tmp/hello-1.0-SNAPSHOT.war
+    deployment: hello.war
+    deployment_strategy: http
+    state: deployed
+    url_username: admin
+    url_password: admin
+
+# Update the hello world application using HTTP management API
+- jboss:
+    src: /tmp/hello-1.0-SNAPSHOT.war
+    deployment: hello.war
+    deployment_strategy: http
+    state: deployed
+    url_username: admin
+    url_password: admin
+
+# Undeploy the hello world application using HTTP management API
+- jboss:
+    deployment: hello.war
+    deployment_strategy: http
+    state: undeployed
+    url_username: admin
+    url_password: admin
+
 """
 
 import os
 import shutil
 import time
+import json
 from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.urls import fetch_url
+
 
+def is_deployed(module):
 
-def is_deployed(deploy_path, deployment):
-    return os.path.exists(os.path.join(deploy_path, "%s.deployed" % deployment))
+    if module.params['deployment_strategy'] == 'http':
+        data_dict = {
+            "operation": "read-resource",
+            "address": {
+                "deployment": module.params['deployment']
+            },
+            "include-runtime": True
+        }
+
+        headers = {
+            'Content-Type': 'application/json'
+        }
+
+        data = json.dumps(data_dict)
+
+        resp, info = fetch_url(module, 'http://%s:9990/management' % module.params['hostname'], data=data, headers=headers)
+
+        return info['status'] == 200
+
+    elif module.params['deployment_strategy'] == 'filesystem':
+        return os.path.exists(os.path.join(module.params['deploy_path'], "%s.deployed" % module.params['deployment']))
+    else:
+        rc, stdout, stderr = module.run_command([module.params['cli_path'], '--connect', '--commands=deploy-info'])
```

Comment:
Some thoughts regarding jboss-cli strategy:
1. `module.run_command()` is being called three times. Maybe it could be refactored to a function with `command(s)` as an argument?
2. While we're at it, wouldn't it be nice to have a module parameter (or a separate module?) that allows execution of arbitrary jboss-cli command(s)?
3. JBoss can be configured to run with a port offset. Should we cover this with a module parameter (mutually exclusive with `port`), or assume that administrator has properly adjusted jboss-cli.xml? This should be documented, whatever decision is taken.
4. If the user executing the module can't [authenticate](https://docs.jboss.org/author/display/WFLY/Command+Line+Interface?_sscc=t#CommandLineInterface-DefaultNativeManagementInterfaceSecurity), we need the username/password. Are they the same as `url_username`/`url_password`? Should we always use them if they are the same?
5. There's no `deploy-info` command on JBoss EAP 6.3.3. But there is `deployment-info`
6. ...and it has a handy argument `--name` that could be used with this value: `module.params['deployment']`
It filters the command's output and when the deployment is present looks like this:
```
NAME                 RUNTIME-NAME         PERSISTENT ENABLED STATUS
foo.ear              foo.ear true         true       OK
```
And like this, when the deployment is not found:
```
JBAS014807: Management resource '[("deployment" => "foo")]' not found
```
Problem with `module.params['deployment'] in line` is that it looks for a substring, so it would be false positive for `"test.ear" in "my-test.ear"`. On the other hand, `deployment-info --name=test.ear` doesn't have that problem.

Cheers,
Marko